nf_core/download.py

```diff
@@ -0,0 +1,196 @@
+
+        # Get workflow details
+        if not self.get_workflow():
+            sys.exit(1)
+
+        # Check that the outdir doesn't already exist
+        if os.path.exists(self.outdir):
+            logging.error("Output directory '{}' already exists".format(self.outdir))
+            sys.exit(1)
+
+        logging.info(
+            "Saving {}".format(self.pipeline) +
+            "\n Pipeline release: {}".format(self.release) +
+            "\n Pull singularity containers: {}".format('Yes' if self.singularity else 'No') +
+            "\n Output directory: {}".format(self.outdir)
+        )
+
+        # Download the pipeline files
+        logging.info("Downloading workflow files from GitHub")
+        self.download_wf_files()
+
+        # Download the singularity images
+        if self.singularity:
+            logging.info("Fetching container names for workflow")
+            self.find_singularity_images()
+            if len(self.containers) == 0:
+                logging.info("No container names found in workflow")
+            else:
+                os.mkdir(os.path.join(self.outdir, 'singularity-images'))
+                for container in self.containers:
+                    self.download_singularity_image(container)
+
+
+    def get_workflow(self):
+        """ Fetch details of nf-core workflow to download """
+        wfs = nf_core.list.Workflows()
+        wfs.get_remote_workflows()
+
+        # Get workflow download details
+        for wf in wfs.remote_workflows:
+            if wf.full_name == self.pipeline or wf.name == self.pipeline:
+
+                # Set pipeline name
+                self.wf_name = wf.name
+
+                # Find latest release hash
+                if self.release is None and len(wf.releases) > 0:
+                    self.release = wf.releases[0]['tag_name']
+                    self.wf_sha = wf.releases[0]['tag_sha']
+                    logging.debug("No release specified. Using latest release: {}".format(self.release))
+                # Find specified release hash
+                elif self.release is not None:
+                    for r in wf.releases:
+                        if r['tag_name'] == self.release.lstrip('v'):
+                            self.wf_sha = r['tag_sha']
+                            break
+                    else:
+                        logging.error("Not able to find release '{}' for {} (found '{}')".format(self.release, wf.full_name, "', '".join([r['tag_name'] for r in wf.releases])))
+                        return False
+                # Must be a dev-only pipeline
+                elif self.release is None:
+                    self.release = 'dev'
+                    self.wf_sha = 'master' # Cheating a little, but GitHub download link works
+                    logging.info("Pipeline is in development. Using current code on master branch.")
+
+                # Set outdir name if not defined
+                if self.outdir is None:
```

Comment:
Maybe this block can be moved up to L40 (keeping all `outdir` related stuff together), so this function will be clean and does only what the name suggests ?

Reply:
I see what you're getting at, but I respectfully disagree 😆 I'm just setting all of the different variables in this function that will be used later on, including the `outdir` and the download link.

Reply:
Hehe, it is a matter of personnel choice/prefference, so no right or wrong here 😇It makes sense in this context to update necessary variables (including `outdir`) in this function as this function was called by `download_workflow` (which uses `outdir` later on).

I suggested this considering more general context i.e. when (or if) `fetch_workflow_details` is called by any other function or script that don't have or need `outdir` variable, then updating `outdir` would be "out of context" 😜(Not an error but its just something not necessary then). That is why I would have put this part up directly in `download_workflow` 😉(Of course, since the class itself is more "download" related this might not happen, but like I said it is matter of personnel preference)